### ISSUE TITLE
Enterprise CSE clusters only provisioning with 2 worker nodes

### ISSUE THREAD

**Issue by mann1mal** [vmware/container-service-extension#377]
I've noticed discrepancies in worker node counts for PKS clusters provisioned via CSE 2.0.0 with respect to the PKS plan configuration assigned to an OvDC. 

Environment:
vCD 9.7
PKS 1.4.0
NSX-T 2.4.1

I have enabled an OvDC for Enterprise CSE with the "small" plan, which was configured in PKS to provision 1 master and 3 workers, all with "small" VM instance profiles.

I provision the cluster via the vcd command below:
~~~
$ vcd cse cluster create prod-cluster
property                     value
---------------------------  ------------------------
kubernetes_master_host       prod-cluster.example.com
kubernetes_master_ips        In Progress
kubernetes_master_port       8443
kubernetes_worker_instances  2  **this is wrong**
last_action                  CREATE
last_action_description      Creating cluster
last_action_state            in progress
name                         prod-cluster
worker_haproxy_ip_addresses
~~~

As you can see, the cluster provisions with 2 workers. I don't have a plan configured in PKS with 2 workers:
~~~
pks plans --json

[
   {
      "id": "8A0E21A8-8072-4D80-B365-D1F502085560",
      "name": "x-small",
      "description": "1 master/1 worker",
      "worker_instances": 1,
      "master_instances": 1,
      "allow-privileged-containers": true
   },
   {
      "id": "58375a45-17f7-4291-acf1-455bfdc8e371",
      "name": "small",
      "description": "1 master/3 workers",
      "worker_instances": 3,
      "master_instances": 1,
      "allow-privileged-containers": true
   },
   {
      "id": "241118e5-69b2-4ef9-b47f-4d2ab071aff5",
      "name": "large",
      "description": "3 masters/3 workers",
      "worker_instances": 3,
      "master_instances": 3,
      "allow-privileged-containers": true
   },
   {
      "id": "2fa824527-318d-4253-9f8e-0025863b8c8a",
      "name": "x-large",
      "description": "3 masters/5 workers",
      "worker_instances": 5,
      "master_instances": 3,
      "allow-privileged-containers": false
   }
]

~~~

If I provision the cluster directly from the PKS cli, I get the expected results:
~~~
# pks create-cluster pksclient-test --external-hostname pksclient-test.example.com --plan small

Name:                     pksclient-test
Plan Name:                small
UUID:                     ee866583-143f-4fa7-ab26-83dbe0715cbe
Last Action:              CREATE
Last Action State:        in progress
Last Action Description:  Creating cluster
Kubernetes Master Host:   pksclient-test.example.com
Kubernetes Master Port:   8443
Worker Nodes:             3 **this is correct**
Kubernetes Master IP(s):  In Progress
Network Profile Name:     
~~~

Both clusters are visible via the PKS cli and are defined with the "small" plan:
~~~
# pks clusters

Name                                                 Plan Name  UUID                                  Status       Action
prod-cluster---52b29a7f-9580-4a61-96d6-fccc8fd37c7c  small      c3d9f2f1-0271-4a35-8e92-950796227605  succeeded    CREATE
pksclient-test                                       small      ee866583-143f-4fa7-ab26-83dbe0715cbe  in progress  CREATE
~~~

I change the plan enabled on the OvDC to the "x-large" plan (3 masters/5 workers):
~~~
$ vcd cse ovdc enable cse-ent-ovdc -o cse-enterprise-org -k ent-pks -p "x-large" -d "example.com"

metadataUpdate: Updating metadata for Virtual Datacenter cse-ent-ovdc(c4928d6f-ffd7-4f29-839b-3915fcc50fc2)
task: b31f229f-c299-4c58-b9c3-e509e99fba16, Updated metadata for Virtual Datacenter cse-ent-ovdc(c4928d6f-ffd7-4f29-839b-3915fcc50fc2), result: success
~~~
And provision a new cluster:
~~~
$ vcd cse cluster create new-prod-cluster
property                     value
---------------------------  ----------------------------
kubernetes_master_host       new-prod-cluster.example.com
kubernetes_master_ips        In Progress
kubernetes_master_port       8443
kubernetes_worker_instances  2 **this is wrong**
last_action                  CREATE
last_action_description      Creating cluster
last_action_state            in progress
name                         new-prod-cluster
worker_haproxy_ip_addresses
~~~
From the PKS cli:
~~~
# pks clusters

Name                                                     Plan Name  UUID                                  Status       Action
new-prod-cluster---52b29a7f-9580-4a61-96d6-fccc8fd37c7c  x-large    acf4b1e3-3726-4a80-a355-c5a735f40450  in progress  CREATE
prod-cluster---52b29a7f-9580-4a61-96d6-fccc8fd37c7c      small      c3d9f2f1-0271-4a35-8e92-950796227605  succeeded    CREATE
pksclient-test                                           small      ee866583-143f-4fa7-ab26-83dbe0715cbe  in progress  CREATE
~~~
It only provisions with 2 worker nodes. I checked the vCenter and 3 masters WERE deployed. So some of the plan information is being considered but not the worker node count config? For what it's worth, I am able to scale the clusters to 3 nodes:
~~~
$ vcd cse cluster resize --nodes=3 prod-cluster
property     value
-----------  ------------
name         prod-cluster
task_status  in progress

$ vcd cse cluster info prod-cluster
property                     value
---------------------------  -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
authorization_mode
compute_profile              {"azs":[{"name":"az-cse-ent-ovdc","cpi":"26167edc19f68246d586","cloud_properties":{"datacenters":[{"name":"Datacenter","clusters":[{"Cluster":{"resource_pool":"cse-ent-ovdc (c4928d6f-ffd7-4f29-839b-3915fcc50fc2)"}}]}]}}],"worker_azs":null,"master_azs":null}
compute_profile_name
kubernetes_master_host       prod-cluster.pks.zpod.io
kubernetes_master_ips        10.96.59.102
kubernetes_master_port       8443
kubernetes_worker_instances  3
last_action                  UPDATE
last_action_description      Instance update in progress
last_action_state            in progress
name                         prod-cluster
network_profile_name
nsxt_network_profile
pks_cluster_name             prod-cluster---52b29a7f-9580-4a61-96d6-fccc8fd37c7c
plan_name                    small
uuid                         c3d9f2f1-0271-4a35-8e92-950796227605
worker_haproxy_ip_addresses
~~~
If I provision an "x-large" cluster from PKS cli, we get the expected results again:
~~~
# pks create-cluster pksclient-test2 --external-hostname pksclient-test2.example.com --plan x-large

# pks cluster pksclient-test2

Name:                     pksclient-test2
Plan Name:                x-large
UUID:                     2e723a58-a654-4e05-aab6-16b5bdaf7101
Last Action:              CREATE
Last Action State:        succeeded
Last Action Description:  Instance provisioning completed
Kubernetes Master Host:   pksclient-test2.example.com
Kubernetes Master Port:   8443
Worker Nodes:             5 **this is correct**
Kubernetes Master IP(s):  10.96.59.104
Network Profile Name:     
~~~

I assumed this was a PKS issue as CSE is just calling whatever plan is presented and PKS does the work but the clusters are provisioning as expected from the PKS cli so not sure where the problem lies. Please let me know if you'd like any log snippets, etc from the environment.

**Comment by mann1mal** [comment]
Just a quick update, upgraded PKS environment to 1.4.1, still same issue:

~~~
$ vcd cse cluster create postupgrade
property                     value
---------------------------  -----------------------
kubernetes_master_host       postupgrade.example.com
kubernetes_master_ips        In Progress
kubernetes_master_port       8443
kubernetes_worker_instances  2 ** this is wrong**
last_action                  CREATE
last_action_description      Creating cluster
last_action_state            in progress
name                         postupgrade
worker_haproxy_ip_addresses
~~~

**Comment by andrew-ni** [comment]
Hi @mann1mal , after some investigation, we think that this is probably a bug in PKS's API. CSE only deals with the plan by name, and never modifies its data. We'll check in with the PKS team.

**Comment by rocknes** [comment]
Hi @mann1mal 

If you take a look at https://github.com/vmware/container-service-extension/blob/ba50ab148ad90c9f44b129a8e4abc4fbdb62f59d/container_service_extension/pksbroker.py#L326
You will see that we are not specifying the number of worker nodes to deploy, we just specify the plan to use while deploying a cluster. I agree with Andrew's assessment that this is probably a PKS api bug. It will be great if you can get in touch with the PKS team regarding this bug.

**Comment by mann1mal** [comment]
Thanks for the response @rocknes and @andrew-ni. I will open a bug against PKS and see what turns up. Thank you!

**Comment by mann1mal** [comment]
One more test I wanted to run, enabled my OvDC with the "x-small" plan (1 master/1 worker), still get 2 worker nodes when I create the cluster:

~~~
$ vcd cse ovdc enable cse-ent-ovdc -o cse-enterprise-org -k ent-pks -p "x-small" -d "example.com"

$ vcd cse cluster create x-small
property                     value
---------------------------  -------------------
kubernetes_master_host       x-small.pks.zpod.io
kubernetes_master_ips        In Progress
kubernetes_master_port       8443
kubernetes_worker_instances  2 **this is wrong**
last_action                  CREATE
last_action_description      Creating cluster
last_action_state            in progress
name                         x-small
worker_haproxy_ip_addresses
~~~

**Comment by tvs** [comment]
> You will see that we are not specifying the number of worker nodes to deploy, we just specify the plan to use while deploying a cluster.

Isn't that what `node_count` is doing here?

https://github.com/vmware/container-service-extension/blob/ba50ab148ad90c9f44b129a8e4abc4fbdb62f59d/container_service_extension/pksbroker.py#L323-L325

**Comment by mann1mal** [comment]
As confirmed by @tvs and @rocknes (offline), the code referenced above with `node_count` variable is what was causing this issue. After making the following change in the `pksbroker.py` script, the clusters provision according to the PKS plan specification, as expected:

replace:
~~~
ClusterParameters(kubernetes_master_host=pks_ext_host,
                              kubernetes_worker_instances=node_count)
~~~

with:
~~~
ClusterParameters(kubernetes_master_host=pks_ext_host)
~~~

Testing the change with the small plan (1 master/3 workers):
~~~
$ vcd cse cluster create demo-cluster
property                     value
---------------------------  ------------------------
kubernetes_master_host       demo-cluster.example.com
kubernetes_master_ips        In Progress
kubernetes_master_port       8443
kubernetes_worker_instances  3 **this is correct**
last_action                  CREATE
last_action_description      Creating cluster
last_action_state            in progress
name                         demo-cluster
worker_haproxy_ip_addresses
~~~

**Comment by andrew-ni** [comment]
This is fixed in #398
